### PR TITLE
Support user-auth service principal

### PR DESCRIPTION
**Pull request by aangelisc**
This PR updates the SDK to support a service principal credential when using user-authentication methods. This will allow us to support backend functionality via the service principal credentials which will always be present.

Changes:
- `ServiceCredentials` and `ServiceCredentialsEnabled` properties added to current user credentials struct
-  Added `UserIdentityFallbackCredentialsEnabled` setting. Enabling this will allow users to specify fallback credentials when creating user identity data sources. This defaults to `true`.
- Added support to user identity auth allowing MSI, Workload Identity, and App Registration credentials to be used as fallback when enabled.
- Added helper functions to determine if a request has been initiated by the Grafana backend, and if the user initiating a request is authenticated with Azure AD.
- Updated `CurrentUserContext` struct and request functions to appropriately set `GrafanaIdToken`
- Added sufficient testing

It should be noted that for backend requests to be identified accurately the feature toggle `idForwarding` should be enabled.

### CODE REVIEW

**Comment by kostrse** [comment]
We need to have design discussion regarding this

File: aztokenprovider/token_provider.go [hunk 1]
```diff
@@ -127,7 +137,14 @@ func (provider *userTokenProvider) GetAccessToken(ctx context.Context, scopes []
 	}
 
 	currentUser, ok := azusercontext.GetCurrentUser(ctx)
-	if !ok {
+	if !ok || currentUser.IdToken == "" && !provider.usernameAssertion {
```
Comment (bossinc) [inline aztokenprovider/token_provider.go:140]: Is this logic right? Should we still try to get the access token if currentUser is ok?

**Review by kostrse — CHANGES_REQUESTED** [review]
AzureSeetings need a new switch to allow/disallow usage of the service credentials.

I would move the fallback logic directly into the user token provider. It needs to be redesigned somehow. Probably should contain an instance of service token provider internally and delegate access token request to it. e.g. work as a composite token provider (a decorator).

File: aztokenprovider/token_provider.go (hunk 1 above)
Comment (aangelisc) [reply aztokenprovider/token_provider.go]: I've reworked this!

File: azcredentials/credentials.go [hunk 2]
```diff
@@ -14,6 +14,7 @@ type AzureCredentials interface {
 
 // AadCurrentUserCredentials "Current User" user identity credentials of the current Grafana user.
 type AadCurrentUserCredentials struct {
+	ServicePrincipal AzureClientSecretCredentials
```
[the remark this answers is not in the document]
Comment (aangelisc) [reply azcredentials/credentials.go]: I've made changes here that should meet the requirements.

File: azsettings/env.go [hunk 3]
```diff
@@ -19,11 +19,12 @@ const (
...
-	UserIdentityTokenURL     = "GFAZPL_USER_IDENTITY_TOKEN_URL"
-	UserIdentityClientID     = "GFAZPL_USER_IDENTITY_CLIENT_ID"
-	UserIdentityClientSecret = "GFAZPL_USER_IDENTITY_CLIENT_SECRET"
-	UserIdentityAssertion    = "GFAZPL_USER_IDENTITY_ASSERTION"
+	UserIdentityEnabled            = "GFAZPL_USER_IDENTITY_ENABLED"
+	UserIdentityTokenURL           = "GFAZPL_USER_IDENTITY_TOKEN_URL"
+	UserIdentityClientID           = "GFAZPL_USER_IDENTITY_CLIENT_ID"
+	UserIdentityClientSecret       = "GFAZPL_USER_IDENTITY_CLIENT_SECRET"
+	UserIdentityAssertion          = "GFAZPL_USER_IDENTITY_ASSERTION"
+	UserIdentityServiceCredentials = "GFAZPL_USER_IDENTITY_SERVICE_CREDENTIALS"
```
Comment (kostrse) [inline azsettings/env.go:27]: Maybe we could come up with a more descriptive name...

File: azsettings/settings.go [hunk 4]
```diff
@@ -16,8 +16,9 @@ type AzureSettings struct {
 	WorkloadIdentityEnabled  bool
 	WorkloadIdentitySettings *WorkloadIdentitySettings
 
-	UserIdentityEnabled       bool
-	UserIdentityTokenEndpoint *TokenEndpointSettings
+	UserIdentityEnabled            bool
+	UserIdentityTokenEndpoint      *TokenEndpointSettings
+	UserIdentityServiceCredentials bool
```
Comment (kostrse) [inline azsettings/settings.go:21]: Let's think about the property name...

Comment (aangelisc) [reply azsettings/settings.go:21]: How about `UserIdentityFallbackCredentialsEnabled`?

**Review by kostrse — DISMISSED** [review]

**Review by asimpson — APPROVED** [review]

**Pull request merged**